qemuarm64-secureboot: Add UEFI Secure Boot v2

jetm opened the pull request:
Hi,

Addressing comments from patch series v1 plus other comments from Mikko Rapeli <mikko.rapeli@linaro.org>.

A backport from meta-ts with the minimal changes to add UEFI Secure Boot into qemuarm64-secureboot machine.

Requirements:

  - Create a UEFI disk partition to copy EFI apps.

  - Add UEFI settings to U-Boot, systemd-boot, and Linux kernel.

  - UEFI keys are to be stored in U-Boot and used to sign systemd-boot and Linux kernel images.

  - Add systemd as Init manager to auto-mount efivarfs.

Introduces uefi-secureboot machine feature.

UEFI keys must be genereated in order to be added to U-Boot. Sign both systemd-boot EFI app and Linux kernel image.

Build and verification steps:

$ kas build 'ci/qemuarm64-secureboot.yml:ci/testimage.yml'

---

Changes since v1:
- Rework all subject commits to follow OE, Yocto, and meta-arm guidelines. 
- Add gen-uefi-sb-keys.bb recipe to generate UEFI keys.
- Add an OE test to validate UEFI Secure Boot.
- Simplify gen_uefi_keys.sh to avoid code repetition.
- Replace grub with systemd-boot.
- Simplify signing binary images with sbsign class.
- Set OE branch to Scarthgap

Changes since the v0:
- Remove u-boot recipe.
- Split the change in several commits.
- Remove sample UEFI keys.
- Validate UEFI keys exist before building.
- Insolate most of changes under uefi-secureboot machine feature.

Commit:
Ignore .vscode

Signed-off-by: Javier Tia <javier.tia@linaro.org>

Commit:
u-boot: Setup UEFI and Secure Boot

Add U-Boot minimal UEFI definitions.

Embedded UEFI variables with the keys previously generated. It's to
enable UEFI Secure Boot and verify the authenticity of the firmware and
operating system.

When U-Boot is built with UEFI support, it includes a set of efivars
that are used to store the Secure Boot variables. These efivars are
embedded in the U-Boot binary and are stored in the flash memory of the
system.

Signed-off-by: Javier Tia <javier.tia@linaro.org>

Commit:
meta-arm: Introduce gen-uefi-sb-keys.bb recipe

Generate a new set of keys on build time. It avoids to use same keys
which could generate a security issue.

Signed-off-by: Javier Tia <javier.tia@linaro.org>

jetm commented on meta-arm-bsp/recipes-bsp/u-boot/u-boot_%.bbappend:
```diff
@@ -5,6 +5,7 @@ MACHINE_U-BOOT_REQUIRE:corstone1000 = "u-boot-corstone1000.inc"
 MACHINE_U-BOOT_REQUIRE:fvp-base = "u-boot-fvp-base.inc"
 MACHINE_U-BOOT_REQUIRE:juno = "u-boot-juno.inc"
 MACHINE_U-BOOT_REQUIRE:tc = "u-boot-tc.inc"
+MACHINE_U-BOOT_REQUIRE:qemuarm64-secureboot = "${@bb.utils.contains('MACHINE_FEATURES', 'uefi-secureboot', 'u-boot-qemuarm64-secureboot.inc', '', d)}"
```
@mikkorapeli-linaro

> I think this should be generic to all machines if uefi-secureboot is in
> MACHINE_FEATURES.
> 
> I know meta-arm will only test qemuarm64-secureboot but users will have different
> machine names and would expect this to work there too.

Make it generic; `uefi-secureboot` would be moved up to another layer, right?

mikkorapeli-linaro replied:
I don't see a need for this to be qemuarm64-secureboot specific. Thus just use a config fragment which is tied to the distro/machine feature, and enable the feature on qemuarm64-secureboot.

jetm replied:
Done

jetm commented on meta-arm/recipes-core/systemd/systemd_%.bbappend:
```diff
@@ -0,0 +1 @@
+require ${@bb.utils.contains('MACHINE_FEATURES', 'uefi-secureboot', 'systemd-uefi-secureboot.inc', '', d)}
```
@mikkorapeli-linaro

> I think this should be machine feature "efi", and this should possibly go to poky
> upstream systemd recipe.
>
> Then, it's possible to switch from grub to systemd-boot with uki binaries
> but that requires some more work. There both kernel and initrd will be signed
> into an efi binary. Only problem for me there is the split between different
> layers: poky, meta-arm, meta-security and meta-secure-core. Not clear which
> bits would belong where, and which layer would contain the test build config
> and oeqa tests.

There are several changes here that are outside the scope of the [initial task](https://linaro.atlassian.net/browse/CAS-236), which does not include the use of uki binaries and systemd-boot. I am not against it, but it's something we will need to discuss further.

And yes, there are also places where these changes should be, because there is already meta-secure-core adding UEFI Secure Boot, but only for x86; ARM is not supported.

jetm commented:
@mikkorapeli-linaro

> Subject should be "linux-yocto: sign kernel image..."

Done. Thanks!

> Could you prefix each patch subject with recipe it actually changes?
>
> That's the pattern in poky and meta-arm. Then if possible, keep
> changes separate to each recipe and main config file (machine,
> kas etc).

Done. Thanks!

> Subject should have "systemd:" prefix.

Done

> Is there some way to test this in oeqa runtime with ssh that
> the boot was really done with secure binaries?
>
> I think this is quite brittle and test should verify that
> boot was secure.

[I have added a test](https://github.com/jetm/meta-arm/pull/1/commits/71dea84f497c027dbe621d4731085a2c1cfaa0a7) to validate UEFI Secure Boot is enabled. Thanks!

mikkorapeli-linaro commented on meta-arm-bsp/recipes-bsp/images/core-image-base-uefi-secureboot.inc:
```diff
@@ -0,0 +1,20 @@
+
+MACHINE_FEATURES += "efi"
+
+EFI_PROVIDER = "grub-efi"
```
If the plan is to use systemd, e.g. DISTRO_FEATURS += "systemd" then this could/should be systemd-boot. That's what poky does already if systemd is the init manager. This would mean that any grub changes could be dropped.

jetm replied:
I have replaced grub with systemd-boot. Thanks!

mikkorapeli-linaro commented on meta-arm/conf/machine/qemuarm64-secureboot.conf:
```diff
@@ -22,3 +22,6 @@ WKS_FILE_DEPENDS = "trusted-firmware-a"
 MACHINE_FEATURES += "optee-ftpm"
+MACHINE_FEATURES += "uefi-secureboot"
+
+INIT_MANAGER = "systemd"
```
Ok, grub support could be dropped completely

jetm replied:
Done!

mikkorapeli-linaro commented on meta-arm-bsp/recipes-bsp/images/core-image-base-uefi-secureboot.inc:
Commit message should have the class name as subject prefix, then maybe a second patch could change the image recipe to use it. In any case, the subject prefix should not be meta-arm.

jetm replied:
Done

mikkorapeli-linaro commented on meta-arm-bsp/recipes-bsp/u-boot/u-boot-qemuarm64-secureboot.inc:
Maybe explain in commit message why the keys are embedded into u-boot efivars. That's to enable the secure boot config with those keys by default.

jetm replied:
Done

mikkorapeli-linaro commented on meta-arm-bsp/recipes-bsp/u-boot/u-boot/qemuarm64-secureboot/qemuarm64-secureboot.cfg:
```diff
@@ -0,0 +1,10 @@
+CONFIG_EFI_SECURE_BOOT=y
+CONFIG_EFI_VARIABLES_PRESEED=y
+CONFIG_PREBOOT="setenv bootmenu_0 UEFI Boot Manager=bootefi bootmgr; setenv bootmenu_1 UEFI Maintenance Menu=eficonfig"
+CONFIG_PREBOOT_DEFINED=y
```
Is anythin in this config snippet actually machine specific? I don't think so. Thus the config fragment name should not be MACHINE specific either. So use uefi-secureboot.cfg which is tied to the distro/machine feature?

jetm replied:
Done

mikkorapeli-linaro commented on ci/qemuarm64-secureboot.yml:
Split the kas and grub changes to separate commits.

Then if target system anyway has systemd, then systemd-boot should be used too and we can drop all changes to grub.

jetm replied:
Done

mikkorapeli-linaro commented on meta-arm/recipes-kernel/linux/linux-yocto-uefi-secureboot.inc:
```diff
@@ -0,0 +1,18 @@
+        "${KERNEL_IMAGE}" \
+        --output "${KERNEL_IMAGETYPE}.signed"
+
+	install -m 0644 "${KERNEL_IMAGETYPE}.signed" "${KERNEL_IMAGE}"
```
I have a sbsign.bbclass to make this easier across recipes. https://gitlab.com/Linaro/trustedsubstrate/meta-ledge-secure/-/blob/systemd_uki/meta-ledge-secure/classes/sbsign.bbclass?ref_type=heads
Not strictly needed but if signature are added to multiple binaries from many recipes then the copy-paste becomes too much.

jetm replied:
class included. Thanks!

mikkorapeli-linaro commented on meta-arm/recipes-core/systemd/systemd-uefi-secureboot.inc:
```diff
@@ -0,0 +1 @@
+PACKAGECONFIG:append = " efi"
```
Based on "efi" distro/machine feature, IMO.

jetm replied:
Don't get this comment. `systemd-uefi-secureboot.inc` is included if `uefi-secureboot` is added to `MACHINE_FEATURES`. [Check it out](https://github.com/jetm/meta-arm/pull/1/files#diff-db8a9a1feea16ab6b5d1ab03e8c595e93b8c4c5e0f94368357aed1ca1bc96725).

jetm replied:
Done

mikkorapeli-linaro commented on meta-arm/lib/oeqa/runtime/cases/uefi_secure_boot.py:
```diff
@@ -0,0 +1,20 @@
+        # Validate Secure Boot is enabled
+        cmd = "efivar -d -n 8be4df61-93ca-11d2-aa0d-00e098032b8c-SecureBoot"
+        status, output = self.target.run(cmd, timeout=120)
+        self.assertEqual(output, '1', msg='\n'.join([cmd, output]))
```
Simple test, thanks! Maybe leave a comment why this single variable is enough.

jetm replied:
Added comment. Thanks

Commit:
qemuarm64-secureboot: Add meta-secure-core layer as dependency

meta-secure-core is required because of sbsigntool.

Signed-off-by: Javier Tia <javier.tia@linaro.org>

Commit:
layer.conf: Introduce UEFI_SB_KEYS_DIR

UEFI_SB_KEYS_DIR saves UEFI keys path.

To avoid security issues, UEFI keys are not provided and they can be
generated by gen_uefi_keys.sh script.

Signed-off-by: Javier Tia <javier.tia@linaro.org>

Commit:
core-image-base: Use UEFI layout disk partitions

- Use efi-disk-no-swap.wks.in disk definition to add expected UEFI disk
  partitions configuration.

Signed-off-by: Javier Tia <javier.tia@linaro.org>

Commit:
linux-yocto: Setup UEFI and sign kernel image

efivarfs kernel module is required to access EFI vars.

Signed-off-by: Javier Tia <javier.tia@linaro.org>

Commit:
systemd: Add UEFI support

Signed-off-by: Javier Tia <javier.tia@linaro.org>

Commit:
meta-arm: Add UEFI Secure Boot test

Add a test to verify UEFI Secure Boot is enabled

Run the test:

kas build 'ci/qemuarm64-secureboot.yml:ci/testimage.yml'

Signed-off-by: Javier Tia <javier.tia@linaro.org>

Commit:
systemd-boot: Use it as bootloader & sign UEFI image

As qemuarm64-secureboot is already using systemd as Init manager, use
too systemd-boot as bootloader. It has a simpler and more intuitive
configuration format compared to grub. It uses a single configuration
file that is easy to understand and modify.

Signed-off-by: Javier Tia <javier.tia@linaro.org>

Commit:
qemuarm64-secureboot: Introduce uefi-secureboot machine feature

Signed-off-by: Javier Tia <javier.tia@linaro.org>

Commit:
uefi-sb-keys.bbclass: Add class to validate UEFI keys

Without UEFI keys, signing will fail and the OS will not boot.

Signed-off-by: Javier Tia <javier.tia@linaro.org>

Commit:
core-image-base: Inherit uefi-sb-keys

Signed-off-by: Javier Tia <javier.tia@linaro.org>

Commit:
sbsign.bbclass: Add class to sign binaries

A lot of recipes are using these same steps to sign binaries
for UEFI secure boot.

Authored-by: Mikko Rapeli <mikko.rapeli@linaro.org>
Signed-off-by: Javier Tia <javier.tia@linaro.org>

jetm commented:
@mikkorapeli-linaro I believe all your comments have been addressed. Please, take another look. Thanks

mikkorapeli-linaro commented on meta-arm/recipes-core/systemd/systemd_%.bbappend:
```diff
@@ -0,0 +1 @@
+require ${@bb.utils.contains('MACHINE_FEATURES', 'efi', 'systemd-uefi-secureboot.inc', '', d)}
```
MACHINE_FEATURE is "efi" but .inc file name still has "uefi-secureboot". Maybe use systemd-efi.inc file name to keep these consistent.

jetm replied:
Good catch. Fixed. Thanks!

mikkorapeli-linaro commented:
One small issue found. With that fixed, please submit to meta-arm. Lets see what they say...

Commit:
qemuarm64-secureboot.yml: Set branch to scarthgap

UEFI Secure Boot is broken in master. Set to the latest stable OE
branch.

Signed-off-by: Javier Tia <javier.tia@linaro.org>

jetm commented:
> One small issue found. With that fixed, please submit to meta-arm. Lets see what they say...

Submitted to meta-arm. I have added another commit to set in kas the release branch to scarthgap because Secure Boot is broken in master. Maybe is the Linux kernel.

jetm commented:
Replaced by https://github.com/jetm/meta-arm/pull/2
